quill cassandra for lagom

WayneWang12 opened the pull request:
### Problem

I'm using Lagom to build my application. And I find that if quill can support lagom then it maybe perfect. So I find a way to add quill support for lagom.

### Solution

Just a little code copy from quill-cassandra and change `Cluster` to lagom's `CassandraSession`.


@jilen Please review it. Thank you~

fwbrasil commented:
@WayneWang12 thank you for the contribution! What's different about lagom? Is it only the driver interface? are the queries different? We should reuse code.

WayneWang12 commented:
@fwbrasil Yes you are right. So I've change my branch. Can you help to review it again? I've extract `quill-cassandra` to `quill-cassandra-core`.

Actually I should create a project for quill-cassandra-akkk. But what I need most now is to support Lagom. So I will come it later to complete the akka part for quill.

WayneWang12 commented:
I'll add unit-test later this day or tommorrow.

jilen commented:
I find it possible to share `com.datastax.driver.core.Session` through `SessionProvider` of akka persistence.

No need to create different context here.

WayneWang12 commented:
It seems I just need `quill-cassandra-core`, then I can extend it to be used by lagom or akka outside quill. 

Whether LagomContext is in quill or not is irrelevent. `CassandraSessionContext` being abstract and independent of `Cluster` makes it convenient to be extended with all kinds of  cassandra drivers. 

@fwbrasil If `quill-cassandra-lagom` is not suitable for quill, I can remove it. Just tell me.

But please extract `quill-cassandra-core` out. That could save me a lot of effort to use quill in Lagom. Thank you!

fwbrasil commented on build.sbt:
```diff
@@ -201,8 +202,8 @@ lazy val `quill-async-postgres` =
 
-lazy val `quill-cassandra` =
-  (project in file("quill-cassandra"))
+lazy val `quill-cassandra-core` =
```
This would make the lagom module also depend on the datastax driver. Let's have the following modules:

1. `quill-cassandra`, which is generic and doesn't have any driver dependency
2. `quill-cassandra-datastax` that depends on `quill-cassandra` and uses the datastax driver
3. `quill-cassandra-lagom` that depends on `quill-cassandra` and uses lagom

Please move as many tests as possible to `quill-cassandra`

WayneWang12 replied:
Hi, I've tried to seperate it. But it seems both `quill-cassandra-datastax` and `quill-cassandra-lagom` will depend on the datastax driver. And It's really hard to seperate it from the common module.

fwbrasil commented on quill-cassandra-core/src/test/cql/cassandra-schema.cql:
```diff
@@ -0,0 +1,198 @@
+CREATE KEYSPACE IF NOT EXISTS quill_test
```
Is this a new file? we don't need to have multiple copies of it since the database can be reused

fwbrasil commented on quill-cassandra-lagom/src/main/scala/io/getquill/CassandraLagomAsyncContext.scala:
```diff
@@ -0,0 +1,64 @@
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+class CassandraLagomAsyncContext[N <: NamingStrategy](
```
There's too much copy-and-paste here. Please create a generic implementation that can be shared between datastax and lagom and only specializes the driver implementation. You could create an interface that needs to be implemented by the driver modules

WayneWang12 commented:
Thanks for your reply. I'm on a business trip last few days in Beijing and back to Shanghai this morning. I'll look into this PR soon after my work.

WayneWang12 commented:
@fwbrasil Seems that I've completed this PR. Can you review it again?

WayneWang12 commented:
Finally, all checks have passed 😂 . It's not easy.

fwbrasil commented on build.sbt:
```diff
@@ -213,15 +213,41 @@ lazy val `quill-cassandra` =
+      libraryDependencies ++= {
+        val lagomVersion = "1.5.0-RC1"
+        Seq(
+          "com.lightbend.lagom" %% "lagom-scaladsl-persistence-cassandra" % lagomVersion % Provided,
```
why `Provided`?

WayneWang12 replied:
Lagom's version keeps changing. I thought it might be better for use to determine their main version.

deusaquilus commented:
... so. We need a `ContextEffect` implementation for:
 * scala.Future - FutureEffect (for CassandraLagomSessionContext and CassandraAsyncContext)
 * akka.stream.scaladsl.Source - AkkaSourceEffect (for the CassandraLagomStreamContext module)
 * monix.reactive.Observable - MonixObserveableEffect (for CassandraStreamContext)
 * monix.eval.Task - Just Reuse 'Runner' (for CassandraMonixContext)

deusaquilus commented:
Here is a sketch what needs to get done: #1322 in order to have modularity (and DRY!) in the Cassandra module. Note how `CassandraAsyncContext` has a completely trivial implementation. `CassandraMonixContext`, and `CassandraStreamContext` (as well as `* CassandraLagomStreamContext` and `* CassandraDatastaxSessionContext` should also be reworked to look like that.

WayneWang12 commented:
@deusaquilus I've read your code. It seems that you want to refactor it to make every effect derived from cluster and session. But Lagom's `CassandraSession` has already done it for us. 

I read Lagom's code carefully this time. They don't expose `Cluster` to outside. It's hard for me to get `Cluster` out. So I want to get rid of `Datasax` and only depend on the core `CassandraSessionContext`. 

Actually every module depends on datasax driver, including `monix`. It's the base driver.

deusaquilus commented:
@WayneWang12 I see what you're saying. I think maybe you should rename `CassandraDatastaxSessionContext` to `CassandraClusteredSessionContext` to indicate the intent.
Also, I don't see the need to introduce a new quill-cassandra-datastax module. It doesn't have any additional dependencies. Other then that I think it's all good.

@fwbrasil Perhaps trying to do a refactor this big is a bit too ambitious for a feature-PR. Do you agree with my general approach of `ContextEffect` though?

WayneWang12 commented:
@deusaquilus At the begining I added a new module named `quill-cassandra-core`. I was meant to extract all the base features into a single module to be depended by `quill-cassandra` and `quill-cassandra-lagom`. I thought it was necessary to hide unneeded classes like `CassandraAsyncContext` as we introduced new `CassandraLagomSessionContext`. That's why I introduce a new module.

But I think my initial `quill-cassandra-core` and `quill-cassandra` modules are better. Because users needn't change their code if they have already used quill for cassandra.  What's your opinion? @deusaquilus

WayneWang12 commented:
@deusaquilus I've updated this PR. And an extra modification is that in `FutureConversion`. I think it may be better if we don't introduce extra executors to convert `ListenableFuture`. Code is as below which I copy from Lagom:
```scala
  implicit class ListenableFutureConverter[A](val lf: ListenableFuture[A]) extends AnyVal {
    def asScala(implicit ec: ExecutionContext): Future[A] = {
      val promise = Promise[A]
      lf.addListener(new Runnable {
        def run(): Unit = {
          promise.complete(Try(lf.get()))
          ()
        }
      }, ec.asInstanceOf[Executor])
      promise.future
    }
  }
```

Previous version will introduce guava's `MoreExecutors` which is a new thread pool. And it have version conflict as I reported in #1293 while this kind haven't.

deusaquilus commented:
@WayneWang12 Looks good! I'll go over it again tomorrow.

deusaquilus commented:
Okay, looks good to me.

deusaquilus commented:
@fwbrasil What do you think? Are you good with this?

WayneWang12 commented:
![image](https://user-images.githubusercontent.com/6812513/52986180-d26e5a00-3431-11e9-90dc-b9d874d05c09.png)
Am I supposed not to click this button??? Sorry I don't know the procedure in Github's new version...

Thanks for your review @deusaquilus ~ Now quill is my first choice for database accessing. Great library you have contributed to the world!

fwbrasil commented on build.sbt:
```diff
@@ -218,11 +218,28 @@ lazy val `quill-cassandra-monix` =
+        )
+      }
+    )
+    .dependsOn(`quill-cassandra` % "compile->compile;test->test")
```
Does the lagom driver depend on the datastax driver? If not we shouldn't have this module depend transitively on the datastax driver

fwbrasil replied:
cool! LGTM

WayneWang12 commented:
Thanks @fwbrasil for accepting my pull request 😄
